WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesAdapter.kt

```diff
@@ -1,121 +1,77 @@
+import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.RecyclerView
-import kotlinx.android.synthetic.main.prepublishing_categories_row.view.*
-import org.apache.commons.text.StringEscapeUtils
-import org.wordpress.android.R
-import org.wordpress.android.models.CategoryNode
-import org.wordpress.android.ui.posts.PrepublishingCategoriesAdapter.CategoriesViewHolder
+import org.wordpress.android.ui.posts.PrepublishingCategoriesViewHolder.PrepublishingCategoriesListItemViewHolder
+import org.wordpress.android.ui.posts.PrepublishingCategoriesViewModel.PrepublishingCategoriesListItemUiState
+import org.wordpress.android.ui.utils.UiHelpers
 
-class PrepublishingCategoriesAdapter(
-    private val context: Context,
-    private val onCheckChangeListener: ((Long, Boolean) -> Unit)
-) : RecyclerView.Adapter<CategoriesViewHolder>() {
-    var categoryNodeList: List<CategoryNode> = arrayListOf()
-        set(value) {
-            if (!isSameCategoryList(value)) {
-                field = value
-                notifyDataSetChanged()
-            }
-        }
-
-    private var selectedCategoryIds: HashSet<Long> = hashSetOf()
-
-    fun set(categoryLevels: List<CategoryNode>, categoriesSelected: HashSet<Long>) {
-        selectedCategoryIds = categoriesSelected
-        categoryNodeList = categoryLevels
-    }
-
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoriesViewHolder {
-        val itemView = LayoutInflater.from(context)
-                .inflate(R.layout.prepublishing_categories_row, parent, false)
-        return CategoriesViewHolder(onCheckChangeListener, itemView)
-    }
-
-    override fun onBindViewHolder(viewHolder: CategoriesViewHolder, position: Int) {
-        viewHolder.bind(categoryNodeList[position], selectedCategoryIds)
-    }
+class PrepublishingCategoriesAdapter(private val uiHelpers: UiHelpers) :
+        RecyclerView.Adapter<PrepublishingCategoriesViewHolder>() {
+    private val items = mutableListOf<PrepublishingCategoriesListItemUiState>()
 
     init {
         setHasStableIds(true)
     }
 
-    override fun getItemId(position: Int): Long {
-        return categoryNodeList[position].categoryId
-    }
-
-    override fun getItemCount(): Int {
-        return categoryNodeList.size
+    override fun onCreateViewHolder(
+        parent: ViewGroup,
+        viewType: Int
+    ): PrepublishingCategoriesViewHolder {
+        return PrepublishingCategoriesListItemViewHolder(parent, uiHelpers)
     }
 
-    private fun isSameCategoryList(categoryList: List<CategoryNode>): Boolean {
-        if (categoryList.size != categoryNodeList.size) {
-            return false
-        }
+    override fun getItemCount(): Int = items.size
 
-        categoryNodeList.forEach {
-            if (!containsNode(it)) {
-                return false
-            }
-        }
-        return true
+    override fun getItemId(position: Int): Long {
+        return position.toLong()
```

Comment:
Is it safe to use position as id? Won't the id/position change when you insert a new item in the list? 

I guess it could be safe as long as we use "hasStableIds = false", just asking to be sure.

Reply:
> I guess it could be safe as long as we use "hasStableIds = false"

That is correct, although it works for both right now, but it is always better to go with false if getItemId is using postion.toLong.

Reply:
> That is correct, although it works for both right now

If it works for both, I guess I don't understand the "hasStableIds" correctly. It shouldn't work when an item changes its id when you change its position, right? I mean it can work in some cases but won't work in other cases when the adapter uses optimization techniques when re-arranging items. Wdyt?

Either way, keeping it as `false` seems like a solution. Thanks